gax-grpc/src/main/java/com/google/api/gax/grpc/OperationPollingCallable.java

```diff
@@ -0,0 +1,125 @@
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.grpc;
+
+import com.google.api.core.ApiClock;
+import com.google.api.core.ApiFuture;
+import com.google.api.core.ApiFutures;
+import com.google.api.gax.retrying.ExponentialRetryAlgorithm;
+import com.google.api.gax.retrying.ResultRetryAlgorithm;
+import com.google.api.gax.retrying.RetrySettings;
+import com.google.api.gax.retrying.TimedAttemptSettings;
+import com.google.longrunning.GetOperationRequest;
+import com.google.longrunning.Operation;
+import com.google.longrunning.OperationsClient;
+import io.grpc.Status;
+import java.util.concurrent.ExecutionException;
+
+/**
+ * A future callable, which relies on callback chaining to execute server polling (asking for a
+ * status of a specific operation: in progress, completed, canceled etc.). This implementation is
+ * called from {@link AttemptCallable}, essentially using for polling the same logic which is used
+ * for retrying.
+ *
+ * @param <RequestT> type of the request
+ */
+class OperationPollingCallable<RequestT> implements FutureCallable<RequestT, Operation> {
+  private final OperationsClient operationsClient;
+  private final ApiFuture<Operation> initialFuture;
+
+  OperationPollingCallable(ApiFuture<Operation> initialFuture, OperationsClient operationsClient) {
+    this.initialFuture = initialFuture;
+    this.operationsClient = operationsClient;
+  }
+
+  /**
+   * This method is supposed to be called from {@link AttemptCallable#call()}
+   *
+   * @param request request
+   * @param context call context
+   */
+  @Override
+  public ApiFuture<Operation> futureCall(RequestT request, CallContext context) {
+    try {
+      if (!initialFuture.isDone() || initialFuture.isCancelled()) {
+        return initialFuture;
+      }
+      // Since initialFuture is done at this point, the following call should be non-blocking
+      Operation initialOperation = initialFuture.get();
+
+      // Note Future.isDone() and Operation.getDone() are two fundamentally different things.
+      if (initialOperation.getDone()) {
+        return initialFuture;
+      }
+
+      GetOperationRequest pollingRequest =
+          GetOperationRequest.newBuilder().setName(initialOperation.getName()).build();
+      return operationsClient.getOperationCallable().futureCall(pollingRequest);
+
+    } catch (ExecutionException e) {
+      return ApiFutures.immediateFailedFuture(e.getCause());
+    } catch (InterruptedException e) {
+      return ApiFutures.immediateFailedFuture(e);
+    }
+  }
+
+  public static class OperationRetryAlgorithm implements ResultRetryAlgorithm<Operation> {
```

Comment:
The two static classes aren't touched by the other code in this file - I suggest lifting them up or putting them somewhere else.